src/overlays/actors/ovl_Bg_Umajump/z_bg_umajump.c

```diff
@@ -27,34 +30,191 @@ const ActorInit Bg_Umajump_InitVars = {
+static InitChainEntry sInitChain[] = {
     ICHAIN_F32(uncullZoneScale, 1200, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneDownward, 300, ICHAIN_CONTINUE),
     ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
 };
 
-#endif
-
-extern InitChainEntry D_8091A810[];
-
-extern UNK_TYPE D_06001220;
-extern UNK_TYPE D_06001558;
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_80919F30.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_80919FC8.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_8091A044.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_8091A0B8.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/BgUmajump_Init.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/BgUmajump_Destroy.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/BgUmajump_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_8091A5A0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Bg_Umajump/func_8091A7B0.s")
+void func_80919F30(BgUmajump* this, PlayState* play) {
+    if (this->rotationTimer >= 90) {
+        this->rotationTimer = 0;
+    }
+
+    this->rotationTimer++;
+    this->dyna.actor.shape.yOffset =
+        Math_SinS((this->rotationTimer / 90.0f) * 65536.0f) * (20.0f / this->dyna.actor.scale.y);
+}
+
+void BgUmajump_StopCutscene(BgUmajump* this, PlayState* play) {
+    if ((play->csCtx.frames >= 6) && !this->hasSoundPlayed) {
+        this->hasSoundPlayed = true;
+        play_sound(NA_SE_EV_KID_HORSE_NEIGH);
+    }
+
+    if (play->csCtx.state == CS_STATE_0) {
+        ActorCutscene_Stop(this->dyna.actor.cutscene);
+        this->dyna.actor.update = Actor_Noop;
+    }
+}
+
+void BgUmajump_PlayCutscene(BgUmajump* this, PlayState* play) {
+    if (ActorCutscene_GetCanPlayNext(this->dyna.actor.cutscene)) {
+        ActorCutscene_StartAndSetUnkLinkFields(this->dyna.actor.cutscene, &this->dyna.actor);
+        gSaveContext.save.weekEventReg[89] |= 0x20;
+        this->actionFunc = BgUmajump_StopCutscene;
+    } else {
+        ActorCutscene_SetIntentToPlay(this->dyna.actor.cutscene);
+    }
+}
+
+void BgUmajump_CheckDistance(BgUmajump* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+
+    if ((this->horse != NULL) && (player != NULL) &&
+        (Math3D_XZDistance(this->horse->world.pos.x, this->horse->world.pos.z, player->actor.world.pos.x,
+                           player->actor.world.pos.z) < 1400.0f)) {
+        this->actionFunc = BgUmajump_PlayCutscene;
+    }
+}
+
+void BgUmajump_Init(Actor* thisx, PlayState* play) {
+    BgUmajump* this = THIS;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+
+    this->actionFunc = NULL;
+    this->horse = NULL;
+    this->hasSoundPlayed = false;
+
+    DynaPolyActor_Init(&this->dyna, 0);
+
+    this->objectIndex = BG_UMAJUMP_GET_OBJECT_INDEX(thisx);
+    thisx->params &= 0xFF;
```

Comment:
Make a macro for this one too. Also if we have to use thisx, we try to use it everywhere else instead of `&this->dyna.actor`, must it only be here?

Reply:
you want me to make a macro for 0xFF? that number alone?
and yes, the thisx thing was something i had to do for it to match, i don't know if it will work everywhere else, i can try...

Reply:
I should clarify, everywhere else in the function, not necessarily the whole actor.

Reply:
> I should clarify, everywhere else in the function, not necessarily the whole actor.

of course